File: apps/web/src/lib/stratos/information-policy.ts
import "server-only";

import { createHash, randomUUID } from "node:crypto";

import { ApiClientError } from "@/lib/types";

export const INFORMATION_POLICY_VERSION = "information-policy-2.0.0";
export const INFORMATION_POLICY_SCHEMA = "stratos-information-policy-2";
export const STRATOS_ORGANIZATION_ID = "org_stratos";

const HANDLING_CLASSES = new Set(["PUBLIC", "INTERNAL", "PROJECT_MANAGEMENT", "RESTRICTED"]);
const TLP_VALUES = new Set(["TLP:RED", "TLP:AMBER+STRICT", "TLP:AMBER", "TLP:GREEN", "TLP:CLEAR"]);
const PAP_VALUES = new Set(["PAP:RED", "PAP:AMBER", "PAP:GREEN", "PAP:CLEAR"]);
const CONTENT_CATEGORIES = new Set([
  "PERSONAL_DATA", "FINANCIAL", "CONTRACTUAL", "SECURITY", "CYBER_THREAT",
  "SOURCE_CODE", "AUTHENTICATION", "AUDIT", "PUBLIC_INFORMATION"
]);
const OBLIGATIONS = new Set([
  "AUDIT_ACCESS", "NO_EXTERNAL_AI", "LOCAL_PROCESSING_ONLY", "NO_PUBLIC_EXPORT",
  "NO_EXPORT", "WATERMARK", "ENCRYPT_AT_REST", "RECIPIENT_CONFIRMATION",
  "ORIGINATOR_APPROVAL", "PAP_ENFORCEMENT"
]);
const SCOPE_TYPES = new Set(["organization", "organization_unit", "budget_scope", "project", "document", "recipient_set", "public"]);

export interface InformationPolicyBinding {
  schemaVersion: typeof INFORMATION_POLICY_SCHEMA;
  policyBindingId: string;
  policyVersion: typeof INFORMATION_POLICY_VERSION;
  handlingClass: "PUBLIC" | "INTERNAL" | "PROJECT_MANAGEMENT" | "RESTRICTED";
  legalClassification: "NONE";
  tlp?: "TLP:RED" | "TLP:AMBER+STRICT" | "TLP:AMBER" | "TLP:GREEN" | "TLP:CLEAR" | null;
  pap?: "PAP:RED" | "PAP:AMBER" | "PAP:GREEN" | "PAP:CLEAR" | null;
  contentCategories: string[];
  audience: {
    organizationId: typeof STRATOS_ORGANIZATION_ID;
    scopeType: "organization" | "organization_unit" | "budget_scope" | "project" | "document" | "recipient_set" | "public";
    scopeIds?: string[];
    recipientSubjectIds?: string[];
  };
  obligations: string[];
  originatorId?: string | null;
  issuedAt: string;
  reviewAt?: string | null;
}

export interface IntegrationEnvelope {
  schemaVersion: "stratos-integration-envelope-1";
  organizationId: typeof STRATOS_ORGANIZATION_ID;
  sourceSystem: "STRATOS_AIIP";
  externalRef: string;
  actor: { type: "person"; subjectId: string };
  sourceResource: {
    governedResourceId: string;
    application: "AIIP";
    resourceType: "idea";
    resourceId: string;
    sourceVersion: string;
    scope:
      | { type: "own"; ownerSubjectId: string }
      | {
          type: "organization" | "organization_unit" | "budget_scope" | "portfolio" | "project" | "document" | "recipient_set";
          id: string;
        };
  };
  correlationId: string;
  idempotencyKey: string;
  policyBindingId: string;
  policyVersion: typeof INFORMATION_POLICY_VERSION;
  policyHash: string;
  classification: {
    handlingClass: string;
    legalClassification: "NONE";
    tlp: string | null;
    pap: string | null;
  };
  payload: {
    operation: "document_upload";
    entityType: "InnovationRequest" | "InnovationRequestImport";
    entityId: string;
    sourceDocumentId: string;
    sha256: string;
  };
}

export interface StratosBudgetGovernanceScope {
  type: "budget_scope";
  id: string;
}

export interface StratosBudgetIntegrationEnvelope {
  schemaVersion: "stratos-integration-envelope-1";
  organizationId: typeof STRATOS_ORGANIZATION_ID;
  sourceSystem: "STRATOS_BUDGET";
  externalRef: string;
  actor: { type: "person"; subjectId: string };
  correlationId: string;
  idempotencyKey: string;
  policyBindingId: string;
  policyVersion: typeof INFORMATION_POLICY_VERSION;
  policyHash: string;
  classification: {
    handlingClass: string;
    legalClassification: "NONE";
    tlp: string | null;
    pap: string | null;
  };
  payload: {
    contractId: string;
    financialScopeKey: string;
    fileHash: string;
  };
}

export function parseInformationPolicy(value: unknown): InformationPolicyBinding {
  const policy = strictRecord(value, "information_policy");
  assertKeys(policy, [
    "schemaVersion", "policyBindingId", "policyVersion", "handlingClass", "legalClassification",
    "tlp", "pap", "contentCategories", "audience", "obligations", "originatorId", "issuedAt", "reviewAt"
  ], "information_policy");
  requiredEqual(policy.schemaVersion, INFORMATION_POLICY_SCHEMA, "schemaVersion");
  requiredEqual(policy.policyVersion, INFORMATION_POLICY_VERSION, "policyVersion");
  requiredEqual(policy.legalClassification, "NONE", "legalClassification", "LEGAL_CLASSIFICATION_UNSUPPORTED");
  const policyBindingId = requiredText(policy.policyBindingId, "policyBindingId");
  if (!/^(?:pol|pb)_[A-Za-z0-9_-]{8,}$/.test(policyBindingId)) fail("POLICY_BINDING_INVALID", "policyBindingId is invalid.");
  const handlingClass = enumText(policy.handlingClass, HANDLING_CLASSES, "handlingClass");
  const tlp = nullableEnum(policy.tlp, TLP_VALUES, "tlp");
  const pap = nullableEnum(policy.pap, PAP_VALUES, "pap");
  const contentCategories = uniqueEnumList(policy.contentCategories, CONTENT_CATEGORIES, "contentCategories");
  const obligations = uniqueEnumList(policy.obligations, OBLIGATIONS, "obligations", "POLICY_OBLIGATION_UNKNOWN");
  const audience = strictRecord(policy.audience, "audience");
  assertKeys(audience, ["organizationId", "scopeType", "scopeIds", "recipientSubjectIds"], "audience");
  requiredEqual(audience.organizationId, STRATOS_ORGANIZATION_ID, "audience.organizationId");
  const scopeType = enumText(audience.scopeType, SCOPE_TYPES, "audience.scopeType");
  const scopeIds = uniqueTextList(audience.scopeIds, "audience.scopeIds");
  const recipientSubjectIds = uniqueTextList(audience.recipientSubjectIds, "audience.recipientSubjectIds");
  const originatorId = optionalText(policy.originatorId);
  if (tlp === "TLP:RED" && (scopeType !== "recipient_set" || recipientSubjectIds.length === 0 || !originatorId)) {
    fail("POLICY_BINDING_INVALID", "TLP:RED requires recipient_set, explicit recipients and originatorId.");
  }
  return {
    schemaVersion: INFORMATION_POLICY_SCHEMA,
    policyBindingId,
    policyVersion: INFORMATION_POLICY_VERSION,
    handlingClass: handlingClass as InformationPolicyBinding["handlingClass"],
    legalClassification: "NONE",
    tlp: tlp as InformationPolicyBinding["tlp"],
    pap: pap as InformationPolicyBinding["pap"],
    contentCategories,
    audience: {
      organizationId: STRATOS_ORGANIZATION_ID,
      scopeType: scopeType as InformationPolicyBinding["audience"]["scopeType"],
      scopeIds,
      recipientSubjectIds
    },
    obligations,
    originatorId,
    issuedAt: requiredTimestamp(policy.issuedAt, "issuedAt"),
    reviewAt: optionalTimestamp(policy.reviewAt, "reviewAt")
  };
}

export function policyHash(policy: InformationPolicyBinding): string {
  return `sha256:${createHash("sha256").update(canonicalJson({
    policyBindingId: policy.policyBindingId,
    policyVersion: policy.policyVersion,
    handlingClass: policy.handlingClass,
    legalClassification: policy.legalClassification,
    tlp: policy.tlp,
    pap: policy.pap,
    obligations: policy.obligations,
    contentCategories: policy.contentCategories,
    audience: policy.audience,
  })).digest("hex")}`;
}

export function createDefaultInformationPolicy(input: {
  classification: "public" | "internal" | "restricted" | "confidential";
  ownerSubjectId: string;
  contentCategories?: string[];
}): InformationPolicyBinding {
  if (input.classification === "confidential") {
    fail("LEGAL_CLASSIFICATION_UNSUPPORTED", "The legacy confidential classification is not supported in Policy V2.");
  }
  const handlingClass = input.classification.toUpperCase() as "PUBLIC" | "INTERNAL" | "RESTRICTED";
  return parseInformationPolicy({
    schemaVersion: INFORMATION_POLICY_SCHEMA,
    policyBindingId: `pol_${randomUUID().replaceAll("-", "")}`,
    policyVersion: INFORMATION_POLICY_VERSION,
    handlingClass,
    legalClassification: "NONE",
    tlp: null,
    pap: null,
    contentCategories: input.contentCategories ?? [],
    audience: {
      organizationId: STRATOS_ORGANIZATION_ID,
      scopeType: "organization",
      scopeIds: [],
      recipientSubjectIds: []
    },
    obligations: handlingClass === "PUBLIC" ? [] : ["AUDIT_ACCESS"],
    originatorId: input.ownerSubjectId,
    issuedAt: new Date().toISOString(),
    reviewAt: null
  });
}

export function parseIntegrationEnvelope(value: unknown, policy: InformationPolicyBinding): IntegrationEnvelope | null {
  if (value === undefined || value === null) return null;
  const envelope = strictRecord(value, "integration_envelope");
  assertKeys(envelope, [
    "schemaVersion", "organizationId", "sourceSystem", "externalRef", "actor", "correlationId",
    "idempotencyKey", "policyBindingId", "policyVersion", "policyHash", "classification", "payload",
    "sourceResource"
  ], "integration_envelope");
  requiredEqual(envelope.schemaVersion, "stratos-integration-envelope-1", "schemaVersion");
  requiredEqual(envelope.organizationId, STRATOS_ORGANIZATION_ID, "organizationId");
  requiredEqual(envelope.sourceSystem, "STRATOS_AIIP", "sourceSystem");
  requiredEqual(envelope.policyVersion, INFORMATION_POLICY_VERSION, "policyVersion");
  requiredEqual(envelope.policyBindingId, policy.policyBindingId, "policyBindingId");
  const authoritativePolicyHash = requiredText(envelope.policyHash, "policyHash").toLowerCase();
  if (!/^sha256:[a-f0-9]{64}$/.test(authoritativePolicyHash)) {
    fail("POLICY_BINDING_INVALID", "policyHash must be a Registry-issued SHA-256 digest.");
  }
  const classification = strictRecord(envelope.classification, "classification");
  assertKeys(classification, ["handlingClass", "legalClassification", "tlp", "pap"], "classification");
  requiredEqual(classification.handlingClass, policy.handlingClass, "classification.handlingClass");
  requiredEqual(classification.legalClassification, "NONE", "classification.legalClassification");
  requiredEqual(classification.tlp ?? null, policy.tlp ?? null, "classification.tlp");
  requiredEqual(classification.pap ?? null, policy.pap ?? null, "classification.pap");
  const actor = strictRecord(envelope.actor, "actor");
  assertKeys(actor, ["type", "subjectId"], "actor");
  requiredEqual(actor.type, "person", "actor.type");
  const sourceResource = strictRecord(envelope.sourceResource, "sourceResource");
  assertKeys(sourceResource, [
    "governedResourceId", "application", "resourceType", "resourceId", "sourceVersion", "scope"
  ], "sourceResource");
  requiredEqual(sourceResource.application, "AIIP", "sourceResource.application");
  requiredEqual(sourceResource.resourceType, "idea", "sourceResource.resourceType");
  const sourceScope = strictRecord(sourceResource.scope, "sourceResource.scope");
  assertKeys(sourceScope, ["type", "id", "ownerSubjectId"], "sourceResource.scope");
  const sourceScopeType = enumText(
    sourceScope.type,
    new Set(["own", "organization", "organization_unit", "budget_scope", "portfolio", "project", "document", "recipient_set"]),
    "sourceResource.scope.type"
  ) as IntegrationEnvelope["sourceResource"]["scope"]["type"];
  const normalizedSourceScope = sourceScopeType === "own"
    ? (() => {
        if (sourceScope.id !== undefined) fail("POLICY_BINDING_INVALID", "An own source scope forbids id.");
        return { type: "own" as const, ownerSubjectId: requiredText(sourceScope.ownerSubjectId, "sourceResource.scope.ownerSubjectId") };
      })()
    : (() => {
        if (sourceScope.ownerSubjectId !== undefined) fail("POLICY_BINDING_INVALID", "ownerSubjectId is valid only for an own source scope.");
        const id = requiredText(sourceScope.id, "sourceResource.scope.id");
        if (sourceScopeType === "organization" && id !== STRATOS_ORGANIZATION_ID) {
          fail("POLICY_BINDING_INVALID", "The organization source scope must identify org_stratos.");
        }
        return { type: sourceScopeType, id };
      })();
  const payload = strictRecord(envelope.payload, "payload");
  assertKeys(payload, ["operation", "entityType", "entityId", "sourceDocumentId", "sha256"], "payload");
  requiredEqual(payload.operation, "document_upload", "payload.operation");
  const entityType = enumText(
    payload.entityType,
    new Set(["InnovationRequest", "InnovationRequestImport"]),
    "payload.entityType"
  ) as IntegrationEnvelope["payload"]["entityType"];
  const payloadSha256 = requiredText(payload.sha256, "payload.sha256").toLowerCase();
  if (!/^sha256:[a-f0-9]{64}$/.test(payloadSha256)) {
    fail("POLICY_BINDING_INVALID", "payload.sha256 must use sha256:<64 lowercase hex chars> format.");
  }
  return {
    schemaVersion: "stratos-integration-envelope-1",
    organizationId: STRATOS_ORGANIZATION_ID,
    sourceSystem: "STRATOS_AIIP",
    externalRef: requiredText(envelope.externalRef, "externalRef"),
    actor: { type: "person", subjectId: requiredText(actor.subjectId, "actor.subjectId") },
    sourceResource: {
      governedResourceId: requiredText(sourceResource.governedResourceId, "sourceResource.governedResourceId"),
      application: "AIIP",
      resourceType: "idea",
      resourceId: requiredText(sourceResource.resourceId, "sourceResource.resourceId"),
      sourceVersion: requiredText(sourceResource.sourceVersion, "sourceResource.sourceVersion"),
      scope: normalizedSourceScope
    },
    correlationId: minText(envelope.correlationId, "correlationId", 8),
    idempotencyKey: minText(envelope.idempotencyKey, "idempotencyKey", 8),
    policyBindingId: policy.policyBindingId,
    policyVersion: INFORMATION_POLICY_VERSION,
    policyHash: authoritativePolicyHash,
    classification: {
      handlingClass: policy.handlingClass,
      legalClassification: "NONE",
      tlp: policy.tlp ?? null,
      pap: policy.pap ?? null
    },
    payload: {
      operation: "document_upload",
      entityType,
      entityId: requiredText(payload.entityId, "payload.entityId"),
      sourceDocumentId: requiredText(payload.sourceDocumentId, "payload.sourceDocumentId"),
      sha256: payloadSha256
    }
  };
}

export function parseStratosBudgetIntegrationEnvelope(
  value: unknown,
  policy: InformationPolicyBinding,
  governanceScopeValue: unknown,
): StratosBudgetIntegrationEnvelope {
  const envelope = strictRecord(value, "integration_envelope");
  assertKeys(envelope, [
    "schemaVersion", "organizationId", "sourceSystem", "externalRef", "actor", "correlationId",
    "idempotencyKey", "policyBindingId", "policyVersion", "policyHash", "classification", "payload"
  ], "integration_envelope");
  requiredEqual(envelope.schemaVersion, "stratos-integration-envelope-1", "schemaVersion");
  requiredEqual(envelope.organizationId, STRATOS_ORGANIZATION_ID, "organizationId");
  requiredEqual(envelope.sourceSystem, "STRATOS_BUDGET", "sourceSystem");
  requiredEqual(envelope.policyVersion, INFORMATION_POLICY_VERSION, "policyVersion");
  requiredEqual(envelope.policyBindingId, policy.policyBindingId, "policyBindingId");

  const authoritativePolicyHash = requiredText(envelope.policyHash, "policyHash").toLowerCase();
  if (!/^sha256:[a-f0-9]{64}$/.test(authoritativePolicyHash)) {
    fail("POLICY_BINDING_INVALID", "policyHash must be a canonical SHA-256 digest.");
  }
  requiredEqual(authoritativePolicyHash, policyHash(policy), "policyHash");

  const classification = strictRecord(envelope.classification, "classification");
  assertKeys(classification, ["handlingClass", "legalClassification", "tlp", "pap"], "classification");
  requiredEqual(classification.handlingClass, policy.handlingClass, "classification.handlingClass");
  requiredEqual(classification.legalClassification, "NONE", "classification.legalClassification");
  requiredEqual(classification.tlp ?? null, policy.tlp ?? null, "classification.tlp");
  requiredEqual(classification.pap ?? null, policy.pap ?? null, "classification.pap");

  const actor = strictRecord(envelope.actor, "actor");
  assertKeys(actor, ["type", "subjectId"], "actor");
  requiredEqual(actor.type, "person", "actor.type");
  const actorSubjectId = requiredText(actor.subjectId, "actor.subjectId");
  if (!/^[A-Za-z0-9][A-Za-z0-9._:@/-]{1,127}$/.test(actorSubjectId)) {
    fail("POLICY_BINDING_INVALID", "actor.subjectId is invalid.");
  }

  const payload = strictRecord(envelope.payload, "payload");
  assertKeys(payload, ["contractId", "financialScopeKey", "fileHash"], "payload");
  const contractId = requiredText(payload.contractId, "payload.contractId");
  if (!/^[A-Za-z0-9][A-Za-z0-9._:-]{0,254}$/.test(contractId)) {
    fail("POLICY_BINDING_INVALID", "payload.contractId is invalid.");
  }
  const financialScopeKey = requiredText(payload.financialScopeKey, "payload.financialScopeKey");
  if (!/^(?:budget-global|budget:[a-z0-9][a-z0-9._-]{0,119})$/.test(financialScopeKey)) {
    fail("POLICY_BINDING_INVALID", "payload.financialScopeKey must identify a canonical budget scope.");
  }
  const fileHash = requiredText(payload.fileHash, "payload.fileHash").toLowerCase();
  if (!/^sha256:[a-f0-9]{64}$/.test(fileHash)) {
    fail("POLICY_BINDING_INVALID", "payload.fileHash must use sha256:<64 lowercase hex chars> format.");
  }

  const externalRef = requiredText(envelope.externalRef, "externalRef");
  const contractExternalRef = `contract:${contractId}`;
  if (
    externalRef !== contractExternalRef
    && !new RegExp(`^${escapeRegex(contractExternalRef)}:document:[a-z0-9][a-z0-9._-]{0,127}$`).test(externalRef)
  ) {
    fail("POLICY_BINDING_INVALID", "externalRef must identify the Budget contract document.");
  }

  const governanceScope = strictRecord(governanceScopeValue, "governance_scope");
  assertKeys(governanceScope, ["type", "id"], "governance_scope");
  requiredEqual(governanceScope.type, "budget_scope", "governance_scope.type");
  requiredEqual(governanceScope.id, financialScopeKey, "governance_scope.id");

  return {
    schemaVersion: "stratos-integration-envelope-1",
    organizationId: STRATOS_ORGANIZATION_ID,
    sourceSystem: "STRATOS_BUDGET",
    externalRef,
    actor: { type: "person", subjectId: actorSubjectId },
    correlationId: minText(envelope.correlationId, "correlationId", 8),
    idempotencyKey: minText(envelope.idempotencyKey, "idempotencyKey", 8),
    policyBindingId: policy.policyBindingId,
    policyVersion: INFORMATION_POLICY_VERSION,
    policyHash: authoritativePolicyHash,
    classification: {
      handlingClass: policy.handlingClass,
      legalClassification: "NONE",
      tlp: policy.tlp ?? null,
      pap: policy.pap ?? null
    },
    payload: { contractId, financialScopeKey, fileHash }
  };
}

function canonicalJson(value: unknown): string {
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  if (value && typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>)
      .filter(([, item]) => item !== undefined)
      .sort(([left], [right]) => left.localeCompare(right));
    return `{${entries.map(([key, item]) => `${JSON.stringify(key)}:${canonicalJson(item)}`).join(",")}}`;
  }
  return JSON.stringify(value);
}

function escapeRegex(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function strictRecord(value: unknown, field: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) fail("POLICY_BINDING_INVALID", `${field} must be an object.`);
  return value as Record<string, unknown>;
}
function assertKeys(value: Record<string, unknown>, keys: string[], field: string) {
  const unknown = Object.keys(value).filter((key) => !keys.includes(key));
  if (unknown.length) fail("POLICY_BINDING_INVALID", `${field} contains unknown fields: ${unknown.join(", ")}.`);
}
function requiredText(value: unknown, field: string): string { if (typeof value !== "string" || !value.trim()) fail("POLICY_BINDING_INVALID", `${field} must be a non-empty string.`); return value.trim(); }
function minText(value: unknown, field: string, length: number): string { const text = requiredText(value, field); if (text.length < length) fail("POLICY_BINDING_INVALID", `${field} is too short.`); return text; }
function optionalText(value: unknown): string | null { if (value === null || value === undefined) return null; if (typeof value !== "string" || !value.trim()) fail("POLICY_BINDING_INVALID", "Optional policy text must be a non-empty string or null."); return value.trim(); }
function optionalTimestamp(value: unknown, field: string): string | null {
  const normalized = optionalText(value);
  if (normalized === null) return null;
  if (!/(?:Z|[+-]\d{2}:\d{2})$/i.test(normalized)) {
    fail("POLICY_BINDING_INVALID", `${field} must include an explicit timezone.`);
  }
  const milliseconds = Date.parse(normalized);
  if (!Number.isFinite(milliseconds)) fail("POLICY_BINDING_INVALID", `${field} must be an ISO-8601 timestamp.`);
  return new Date(milliseconds).toISOString();
}
function requiredTimestamp(value: unknown, field: string): string {
  const normalized = optionalTimestamp(value, field);
  if (normalized === null) fail("POLICY_BINDING_INVALID", `${field} is required.`);
  return normalized;
}
function requiredEqual(value: unknown, expected: unknown, field: string, code = "POLICY_BINDING_INVALID") { if (value !== expected) fail(code, `${field} is not supported.`); }
function enumText(value: unknown, allowed: Set<string>, field: string): string { const text = requiredText(value, field); if (!allowed.has(text)) fail("POLICY_BINDING_INVALID", `${field} is unknown.`); return text; }
function nullableEnum(value: unknown, allowed: Set<string>, field: string): string | null { if (value === null || value === undefined) return null; return enumText(value, allowed, field); }
function uniqueTextList(value: unknown, field: string): string[] { if (value === undefined) return []; if (!Array.isArray(value)) fail("POLICY_BINDING_INVALID", `${field} must be an array.`); const items = value.map((item) => requiredText(item, field)); if (new Set(items).size !== items.length) fail("POLICY_BINDING_INVALID", `${field} must be unique.`); return items; }
function uniqueEnumList(value: unknown, allowed: Set<string>, field: string, code = "POLICY_BINDING_INVALID"): string[] { const items = uniqueTextList(value, field); const unknown = items.filter((item) => !allowed.has(item)); if (unknown.length) fail(code, `${field} contains unknown values.`); return items; }
function fail(code: string, message: string): never { throw new ApiClientError(message, 422, code, "web-information-policy"); }

File: services/registry-api/app/information_policy.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from hashlib import sha256
import json
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


POLICY_SCHEMA_VERSION = "stratos-information-policy-2"
POLICY_VERSION = "information-policy-2.0.0"
ORGANIZATION_ID = "org_stratos"


class HandlingClass(str, Enum):
    public = "PUBLIC"
    internal = "INTERNAL"
    restricted = "RESTRICTED"


class TlpLabel(str, Enum):
    red = "TLP:RED"
    amber_strict = "TLP:AMBER+STRICT"
    amber = "TLP:AMBER"
    green = "TLP:GREEN"
    clear = "TLP:CLEAR"


class PapLabel(str, Enum):
    red = "PAP:RED"
    amber = "PAP:AMBER"
    green = "PAP:GREEN"
    clear = "PAP:CLEAR"


class ContentCategory(str, Enum):
    personal_data = "PERSONAL_DATA"
    financial = "FINANCIAL"
    contractual = "CONTRACTUAL"
    security = "SECURITY"
    cyber_threat = "CYBER_THREAT"
    source_code = "SOURCE_CODE"
    authentication = "AUTHENTICATION"
    audit = "AUDIT"
    public_information = "PUBLIC_INFORMATION"


class PolicyObligation(str, Enum):
    audit_access = "AUDIT_ACCESS"
    no_external_ai = "NO_EXTERNAL_AI"
    local_processing_only = "LOCAL_PROCESSING_ONLY"
    no_public_export = "NO_PUBLIC_EXPORT"
    no_export = "NO_EXPORT"
    watermark = "WATERMARK"
    encrypt_at_rest = "ENCRYPT_AT_REST"
    recipient_confirmation = "RECIPIENT_CONFIRMATION"
    originator_approval = "ORIGINATOR_APPROVAL"
    pap_enforcement = "PAP_ENFORCEMENT"


class PolicyAudience(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    organization_id: Literal["org_stratos"] = Field(alias="organizationId")
    scope_type: Literal[
        "organization", "organization_unit", "project", "document", "recipient_set", "public"
    ] = Field(alias="scopeType")
    scope_ids: list[str] = Field(default_factory=list, alias="scopeIds")
    recipient_subject_ids: list[str] = Field(default_factory=list, alias="recipientSubjectIds")


class InformationPolicyBinding(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True, use_enum_values=True)

    schema_version: Literal["stratos-information-policy-2"] = Field(alias="schemaVersion")
    policy_binding_id: str = Field(alias="policyBindingId", pattern=r"^(?:pol|pb)_[A-Za-z0-9_-]{8,}$")
    policy_version: Literal["information-policy-2.0.0"] = Field(alias="policyVersion")
    handling_class: HandlingClass = Field(alias="handlingClass")
    legal_classification: Literal["NONE"] = Field(alias="legalClassification")
    tlp: TlpLabel | None = None
    pap: PapLabel | None = None
    content_categories: list[ContentCategory] = Field(alias="contentCategories")
    audience: PolicyAudience
    obligations: list[PolicyObligation]
    originator_id: str | None = Field(default=None, alias="originatorId")
    issued_at: datetime | None = Field(default=None, alias="issuedAt")
    review_at: datetime | None = Field(default=None, alias="reviewAt")

    @model_validator(mode="after")
    def validate_semantics(self) -> "InformationPolicyBinding":
        if len(set(self.content_categories)) != len(self.content_categories):
            raise ValueError("contentCategories must contain unique values")
        if len(set(self.obligations)) != len(self.obligations):
            raise ValueError("obligations must contain unique values")
        if len(set(self.audience.scope_ids)) != len(self.audience.scope_ids):
            raise ValueError("audience.scopeIds must contain unique values")
        if len(set(self.audience.recipient_subject_ids)) != len(
            self.audience.recipient_subject_ids
        ):
            raise ValueError("audience.recipientSubjectIds must contain unique values")
        if self.tlp == TlpLabel.red.value:
            if self.audience.scope_type != "recipient_set":
                raise ValueError("TLP:RED requires audience.scopeType=recipient_set")
            if not self.audience.recipient_subject_ids:
                raise ValueError("TLP:RED requires explicit recipientSubjectIds")
            if not self.originator_id:
                raise ValueError("TLP:RED requires originatorId")
        return self


class IntegrationActor(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal["person", "service"]
    subject_id: str = Field(alias="subjectId", min_length=1)


class IntegrationGovernanceScope(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    type: Literal[
        "own",
        "organization",
        "organization_unit",
        "budget_scope",
        "portfolio",
        "project",
        "document",
        "recipient_set",
    ]
    id: str | None = Field(default=None, min_length=1, max_length=160)
    owner_subject_id: str | None = Field(
        default=None,
        alias="ownerSubjectId",
        min_length=1,
        max_length=160,
    )

    @model_validator(mode="after")
    def validate_shape(self) -> "IntegrationGovernanceScope":
        if self.type == "own":
            if self.id is not None or not self.owner_subject_id:
                raise ValueError("An own scope requires ownerSubjectId and forbids id")
            return self
        if self.owner_subject_id is not None:
            raise ValueError("ownerSubjectId is valid only for an own scope")
        if self.type == "organization":
            if self.id != "org_stratos":
                raise ValueError("The organization scope must identify org_stratos")
            return self
        if not self.id:
            raise ValueError("A non-organization scope requires id")
        return self


class IntegrationSourceResource(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    governed_resource_id: str = Field(
        alias="governedResourceId", min_length=1, max_length=160
    )
    application: Literal["AIIP"]
    resource_type: Literal["idea"] = Field(alias="resourceType")
    resource_id: str = Field(alias="resourceId", min_length=1, max_length=160)
    source_version: str = Field(alias="sourceVersion", min_length=1, max_length=160)
    scope: IntegrationGovernanceScope


class IntegrationClassification(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True, use_enum_values=True)

    handling_class: HandlingClass = Field(alias="handlingClass")
    legal_classification: Literal["NONE"] = Field(alias="legalClassification")
    tlp: TlpLabel | None
    pap: PapLabel | None


class IntegrationEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True, use_enum_values=True)

    schema_version: Literal["stratos-integration-envelope-1"] = Field(alias="schemaVersion")
    organization_id: Literal["org_stratos"] = Field(alias="organizationId")
    source_system: Literal[
        "STRATOS_AIIP",
        "STRATOS_ARCHFLOW",
        "STRATOS_BUDGET",
        "STRATOS_PROJECTFLOW",
        "STRATOS_AKB",
        "STRATOS_SECURITY_PREFLIGHT",
    ] = Field(alias="sourceSystem")
    external_ref: str = Field(alias="externalRef", min_length=1, max_length=300)
    actor: IntegrationActor
    source_resource: IntegrationSourceResource | None = Field(
        default=None, alias="sourceResource"
    )
    correlation_id: str = Field(alias="correlationId", min_length=8, max_length=200)
    idempotency_key: str = Field(alias="idempotencyKey", min_length=8, max_length=200)
    policy_binding_id: str = Field(alias="policyBindingId", min_length=8)
    policy_version: Literal["information-policy-2.0.0"] = Field(alias="policyVersion")
    policy_hash: str = Field(alias="policyHash", pattern=r"^sha256:[a-f0-9]{64}$")
    classification: IntegrationClassification
    payload: dict[str, Any]


class AiipUploadActor(IntegrationActor):
    type: Literal["person"]


class AiipUploadEnvelopePayload(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    operation: Literal["document_upload"]
    entity_type: Literal["InnovationRequest", "InnovationRequestImport"] = Field(
        alias="entityType"
    )
    entity_id: str = Field(alias="entityId", min_length=1, max_length=160)
    source_document_id: str = Field(
        alias="sourceDocumentId", min_length=1, max_length=1024
    )
    sha256: str = Field(pattern=r"^sha256:[a-f0-9]{64}$")


class AiipUploadIntegrationEnvelope(IntegrationEnvelope):
    """Exact immutable lineage envelope accepted by the dedicated AIIP route."""

    source_system: Literal["STRATOS_AIIP"] = Field(alias="sourceSystem")
    actor: AiipUploadActor
    source_resource: IntegrationSourceResource = Field(alias="sourceResource")
    payload: AiipUploadEnvelopePayload


def canonical_policy_payload(binding: InformationPolicyBinding) -> dict[str, Any]:
    value = binding.model_dump(mode="json", by_alias=True, exclude_none=False)
    return {
        "policyBindingId": value["policyBindingId"],
        "policyVersion": value["policyVersion"],
        "handlingClass": value["handlingClass"],
        "legalClassification": value["legalClassification"],
        "tlp": value["tlp"],
        "pap": value["pap"],
        "obligations": value["obligations"],
        "contentCategories": value["contentCategories"],
        "audience": value["audience"],
    }


def canonical_policy_hash(binding: InformationPolicyBinding) -> str:
    encoded = json.dumps(
        canonical_policy_payload(binding),
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")
    return f"sha256:{sha256(encoded).hexdigest()}"


def policy_columns(binding: InformationPolicyBinding | None) -> dict[str, Any]:
    if binding is None:
        return {
            "organization_id": ORGANIZATION_ID,
            "policy_binding_id": None,
            "policy_version": None,
            "policy_hash": None,
            "policy_summary": {},
        }
    return {
        "organization_id": binding.audience.organization_id,
        "policy_binding_id": binding.policy_binding_id,
        "policy_version": binding.policy_version,
        "policy_hash": canonical_policy_hash(binding),
        "policy_summary": binding.model_dump(mode="json", by_alias=True, exclude_none=False),
    }


def legacy_classification(binding: InformationPolicyBinding) -> str:
    return {
        HandlingClass.public.value: "public",
        HandlingClass.internal.value: "internal",
        HandlingClass.restricted.value: "restricted",
    }[str(binding.handling_class)]


def anonymous_public_eligible(binding: InformationPolicyBinding) -> bool:
    """Return true only for a policy that may enter STRATOS public-publication approval.

    This does not publish content and never substitutes for an active central
    InformationPublication plus a fresh public decision.
    """

    categories = set(binding.content_categories)
    obligations = set(binding.obligations)
    return bool(
        binding.handling_class == HandlingClass.public.value
        and binding.legal_classification == "NONE"
        and binding.audience.scope_type == "public"
        and not binding.audience.scope_ids
        and not binding.audience.recipient_subject_ids
        and binding.tlp in {None, TlpLabel.clear.value}
        and binding.pap in {None, PapLabel.clear.value}
        and ContentCategory.public_information.value in categories
        and not {ContentCategory.personal_data.value, ContentCategory.authentication.value}.intersection(categories)
        and PolicyObligation.audit_access.value in obligations
        and not {
            PolicyObligation.no_export.value,
            PolicyObligation.no_public_export.value,
        }.intersection(obligations)
    )
